File: src/test/scala/com/exasol/cloudetl/bucket/GCSBucketTest.scala
package com.exasol.cloudetl.bucket

import java.nio.file.Files
import java.nio.file.Path

import org.scalatest.prop.TableDrivenPropertyChecks._

import scala.jdk.CollectionConverters.MapHasAsScala

class GCSBucketTest extends AbstractBucketTest {

  private[this] val defaultProperties = Map(
    PATH -> "gs://my-bucket/",
    FORMAT -> "AVRO",
    "GCS_KEYFILE_PATH" -> "/keyfile.json",
    "GCS_PROJECT_ID" -> "myProject"
  )

  test("constructs bucket properly") {
    val bucket = getBucket(defaultProperties)
    bucket.validate()
    assert(bucket.isInstanceOf[GCSBucket])
  }

  test("validation fails when both keyfile and connection are specified") {
    val bucket = getBucket(
      Map(
        PATH -> "gs://my-bucket/",
        FORMAT -> "AVRO",
        "GCS_PROJECT_ID" -> "myProject",
        "GCS_KEYFILE_PATH" -> "/keyfile.json",
        "CONNECTION_NAME" -> "GCS_CONNECTION"
      )
    )
    val thrown = intercept[IllegalArgumentException] {
      bucket.validate()
    }
    assert(
      thrown.getMessage() === "E-CSE-30: Both properties 'GCS_KEYFILE_PATH' and 'CONNECTION_NAME' are specified. "
        + "Please specify only one of them."
    )
  }

  test("validation fails when both keyfile and connection are missing") {
    val bucket = getBucket(
      Map(
        PATH -> "gs://my-bucket/",
        FORMAT -> "AVRO",
        "GCS_PROJECT_ID" -> "myProject"
      )
    )
    val thrown = intercept[IllegalArgumentException] {
      bucket.validate()
    }
    assert(
      thrown.getMessage() === "E-CSE-31: Neither of properties 'GCS_KEYFILE_PATH' or 'CONNECTION_NAME' is specified. "
        + "Please specify exactly one of them."
    )
  }

  val invalidConnectionContent = Table(
    (
      "",
      "E-IEUCS-4: Properties input string does not contain key-value assignment '='. "
        + "Please make sure that key-value pairs encoded correctly."
    ),
    (
      "wrong-format",
      "E-IEUCS-4: Properties input string does not contain key-value assignment '='. "
        + "Please make sure that key-value pairs encoded correctly."
    ),
    (
      "wrong_key=value",
      "E-CSE-32: The connection 'connection_info' does not contain 'GCS_KEYFILE_CONTENT' property. "
        + "Please check the connection properties."
    ),
    (
      "GCS_KEYFILE_CONTENT=invalid_json",
      "E-CSE-33: The connection 'connection_info' does not contain valid JSON in property 'GCS_KEYFILE_CONTENT'. "
        + "Please check the connection properties."
    )
  )
  forAll(invalidConnectionContent) { (connectionContent: String, expectedErrorMessage: String) =>
    test(s"validation fails for invalid connection value '$connectionContent'") {
      val bucket = getBucket(
        Map(
          PATH -> "gs://my-bucket/",
          FORMAT -> "AVRO",
          "GCS_PROJECT_ID" -> "myProject",
          "CONNECTION_NAME" -> "connection_info"
        ),
        mockConnectionInfo("", connectionContent)
      )
      val thrown = intercept[IllegalArgumentException] {
        bucket.validate()
      }
      assert(
        thrown.getMessage() === expectedErrorMessage
      )
    }
  }

  test("keyfile path configured when path property specified") {
    val bucket = getBucket(defaultProperties)
    bucket.validate()
    assert(bucket.isInstanceOf[GCSBucket])
    val conf = bucket.getConfiguration()
    assert(conf.get("fs.gs.auth.service.account.json.keyfile") === "/keyfile.json")
  }

  test("keyfile path configured when connection specified") {
    val bucket = getBucket(
      Map(
        PATH -> "gs://my-bucket/",
        FORMAT -> "AVRO",
        "GCS_PROJECT_ID" -> "myProject",
        "CONNECTION_NAME" -> "connection_info"
      ),
      mockConnectionInfo("", "GCS_KEYFILE_CONTENT={\"key\":\"value\"}")
    )
    bucket.validate()
    assert(bucket.isInstanceOf[GCSBucket])
    val conf = bucket.getConfiguration()
    val tempFilePath = conf.get("fs.gs.auth.service.account.json.keyfile")
    assert(!tempFilePath.isBlank())
    assert(Files.readString(Path.of(tempFilePath)) === "{\"key\":\"value\"}")
  }

  test("proxy settings should be added when present") {
    val bucket = getBucket(
      defaultProperties ++ Map(
        "PROXY_HOST" -> "myproxy.net",
        "PROXY_PORT" -> "3198",
        "PROXY_USERNAME" -> "user",
        "PROXY_PASSWORD" -> "password"
      )
    )
    bucket.validate()
    assert(bucket.isInstanceOf[GCSBucket])
    val conf = bucket.getConfiguration()
    assert(conf.get("fs.gs.proxy.address") === "myproxy.net:3198")
    assert(conf.get("fs.gs.proxy.username") === "user")
    assert(conf.get("fs.gs.proxy.password") === "password")
  }

  test("proxy settings should not be added without host") {
    val bucket = getBucket(
      defaultProperties ++ Map(
        "PROXY_PORT" -> "3198",
        "PROXY_USERNAME" -> "user",
        "PROXY_PASSWORD" -> "password"
      )
    )
    bucket.validate()
    assert(bucket.isInstanceOf[GCSBucket])
    val conf = bucket.getConfiguration()
    assert(conf.getValByRegex("fs\\.s3a\\.proxy.+").asScala === Map.empty[String, String])
  }
}

File: src/test/java/com/exasol/cloudetl/bucket/GCSBucketTest.java
package com.exasol.cloudetl.bucket;

import static com.exasol.cloudetl.bucket.AzureAbfsBucketTest.withAll;
import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.stream.Stream;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class GCSBucketTest extends AbstractBucketTest {
    private final Map<String, String> defaultProperties = Map.of(PATH, "gs://my-bucket/", FORMAT, "AVRO",
            "GCS_KEYFILE_PATH", "/keyfile.json", "GCS_PROJECT_ID", "myProject");

    @Test
    void constructsBucketProperly() {
        final Bucket bucket = getBucket(this.defaultProperties);
        bucket.validate();
        assertInstanceOf(GCSBucket.class, bucket);
    }

    @Test
    void validationFailsWhenBothKeyfileAndConnectionAreSpecified() {
        final Bucket bucket = getBucket(Map.of(PATH, "gs://my-bucket/", FORMAT, "AVRO", "GCS_PROJECT_ID",
                "myProject", "GCS_KEYFILE_PATH", "/keyfile.json", "CONNECTION_NAME", "GCS_CONNECTION"));
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, bucket::validate);
        assertEquals("E-CSE-30: Both properties 'GCS_KEYFILE_PATH' and 'CONNECTION_NAME' are specified. "
                + "Please specify only one of them.", thrown.getMessage());
    }

    @Test
    void validationFailsWhenBothKeyfileAndConnectionAreMissing() {
        final Bucket bucket = getBucket(Map.of(PATH, "gs://my-bucket/", FORMAT, "AVRO", "GCS_PROJECT_ID", "myProject"));
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, bucket::validate);
        assertEquals("E-CSE-31: Neither of properties 'GCS_KEYFILE_PATH' or 'CONNECTION_NAME' is specified. "
                + "Please specify exactly one of them.", thrown.getMessage());
    }

    @ParameterizedTest
    @MethodSource("invalidConnectionValues")
    void validationFailsForInvalidConnectionValues(final String connectionContent, final String expectedErrorMessage) {
        final Bucket bucket = getBucket(Map.of(PATH, "gs://my-bucket/", FORMAT, "AVRO", "GCS_PROJECT_ID",
                "myProject", "CONNECTION_NAME", "connection_info"), mockConnectionInfo("", connectionContent));
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, bucket::validate);
        assertEquals(expectedErrorMessage, thrown.getMessage());
    }

    @Test
    void keyfilePathConfiguredWhenPathPropertySpecified() {
        final Bucket bucket = getBucket(this.defaultProperties);
        bucket.validate();
        assertInstanceOf(GCSBucket.class, bucket);
        assertEquals("/keyfile.json", bucket.getConfiguration().get("fs.gs.auth.service.account.json.keyfile"));
    }

    @Test
    void keyfilePathConfiguredWhenConnectionSpecified() throws IOException {
        final Bucket bucket = getBucket(Map.of(PATH, "gs://my-bucket/", FORMAT, "AVRO", "GCS_PROJECT_ID",
                "myProject", "CONNECTION_NAME", "connection_info"),
                mockConnectionInfo("", "GCS_KEYFILE_CONTENT={\"key\":\"value\"}"));
        bucket.validate();
        assertInstanceOf(GCSBucket.class, bucket);
        final String tempFilePath = bucket.getConfiguration().get("fs.gs.auth.service.account.json.keyfile");
        assertFalse(tempFilePath.isBlank());
        assertEquals("{\"key\":\"value\"}", Files.readString(Path.of(tempFilePath)));
    }

    @Test
    void proxySettingsShouldBeAddedWhenPresent() {
        final Bucket bucket = getBucket(withAll(this.defaultProperties, Map.of("PROXY_HOST", "myproxy.net",
                "PROXY_PORT", "3198", "PROXY_USERNAME", "user", "PROXY_PASSWORD", "password")));
        bucket.validate();
        assertInstanceOf(GCSBucket.class, bucket);
        final org.apache.hadoop.conf.Configuration conf = bucket.getConfiguration();
        assertEquals("myproxy.net:3198", conf.get("fs.gs.proxy.address"));
        assertEquals("user", conf.get("fs.gs.proxy.username"));
        assertEquals("password", conf.get("fs.gs.proxy.password"));
    }

    @Test
    void proxySettingsShouldNotBeAddedWithoutHost() {
        final Bucket bucket = getBucket(withAll(this.defaultProperties,
                Map.of("PROXY_PORT", "3198", "PROXY_USERNAME", "user", "PROXY_PASSWORD", "password")));
        bucket.validate();
        assertInstanceOf(GCSBucket.class, bucket);
        assertTrue(bucket.getConfiguration().getValByRegex("fs\\.s3a\\.proxy.+").isEmpty());
    }

    private static Stream<Arguments> invalidConnectionValues() {
        final String invalidFormatMessage = "E-IEUCS-4: Properties input string does not contain key-value assignment '='."
                + " Please make sure that key-value pairs encoded correctly.";
        return Stream.of(//
                Arguments.of("", invalidFormatMessage), //
                Arguments.of("wrong-format", invalidFormatMessage), //
                Arguments.of("wrong_key=value",
                        "E-CSE-32: The connection 'connection_info' does not contain 'GCS_KEYFILE_CONTENT' property."
                                + " Please check the connection properties."), //
                Arguments.of("GCS_KEYFILE_CONTENT=invalid_json",
                        "E-CSE-33: The connection 'connection_info' does not contain valid JSON in property"
                                + " 'GCS_KEYFILE_CONTENT'. Please check the connection properties."));
    }
}
